keps/sig-apps/20190715-rollingUpdate-enhancement-for-daemonset.md

```diff
@@ -0,0 +1,182 @@
+---
+title: RollingUpdate enhancement for Daemonset
+authors:
+  - "@resouer"
+  - "@zhangxiaoyu-zidif"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-apps
+reviewers:
+  - "@janetkuo"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-07-15
+last-updated: 2019-11-13
+status: provisional
+see-also:
+  - n/a
+replaces:
+  - n/a
+superseded-by:
+  - n/a
+---
+
+# RollingUpdate enhancement for Daemonset
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+  - [Implementation Details](#implementation-details)
+    - [API Changes](#api-changes)
+    - [Implementation](#implementation)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Upgrades/Downgrades](#upgradesdowngrades)
+  - [Tests](#tests)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+The purpose of this enhancement is to enhance RollingUpdate for Daemonset.
+  - When Daemonset rolling update, it can be `Paused` by user.
+  - Add `SurgingRollingUpdate` for Daemonset as a new update strategy.
+  - Add two more fields, `partition` and `selector` for `RollingUpdate`
+    - `partition` is the number of DaemonSet pods remained to be old version.
+    - `selector` is to query over nodes whoes labels are matched by the Daemonset `RollingUpdate`.
+
+
+## Motivation
+
+Consider the following scenarios:-
+
+1. When executing Daemonset update, even those Pods run well, user need to pause updation to check if  some results of new version image meet expect. If not, user can rollback images to stable version.
+
+1. When rollingUpdate, user should have some chioces to select some specific or random nodes to update pods running on it.
```

Comment:
Or, we can control the progress of the upgrade, such as setting the target percentage of the upgrade.

Reply:
yes, that's why we want to use field `partition` enhance update strategy.

Reply:
This is really essential in a complex private cloud environment.

Reply:
We are also landing DaemonSet to upgrade some agent components. We also encountered these problems. From the current point of view, the only feasible solution is CRD. Looking forward to this KEP landing.

Reply:
Yes, we will send our implement to other project to release this enhancement at the same time. But if it could be adopted by K/K, that is the best choice.

Reply:
Cool, agree with U.